Python/specialize.c

```diff
@@ -2414,6 +2414,35 @@ binary_op_fail_kind(int oparg, PyObject *lhs, PyObject *rhs)
 
 /** Binary Op Specialization Extensions */
 
+/* long-long */
+
+static inline int
+is_nonnegative_compactlong(PyObject *v)
+{
+    return PyLong_CheckExact(v) &&
+           (!_PyLong_IsNegative((PyLongObject *)v)) &&
+           _PyLong_IsCompact((PyLongObject *)v);
+}
+
+static int
+nonnegative_compactlongs_guard(PyObject *lhs, PyObject *rhs)
+{
+    return (is_nonnegative_compactlong(lhs) && is_nonnegative_compactlong(rhs));
+}
+
+#define NONNEGATIVE_LONGS_ACTION(NAME, OP) \
```

Comment:
Why restrict to nonnegative longs here? If we do restrict, then we can replace the calls _PyLong_CompactValue with direct access to `op->long_value.ob_digit[0]`

Reply:
Because negative ints need more work at runtime and I don't think they're common with bitwise logical ops.

Reply:
The extra work is already done by `_PyLong_CompactValue` or am missing something? The output of `ls OP rhs` might not be a compact int, but there are no guards for the output type.